### ISSUE TITLE
zenoh plugins statically linked in a zenoh app

### ISSUE THREAD

**Issue by JEnoch** [eclipse-zenoh/zenoh#89]
Currently (v0.5.0-beta.8) the zenoh plugins are dynamic libraries that are loaded by `zenohd` at startup.
They just must implement 2 `#[no_mangle]` functions:
```rust
// return the list of command lines arguments expected by the plugin
fn get_expected_args<'a, 'b>() -> Vec<clap::Arg<'a, 'b>>;

// start the plugin with command line arguments
fn start(runtime: Runtime, args: clap::ArgMatches<'static>);
```

Now we would like to be able to use the zenoh plugins in any zenoh Rust application, but in a static way (i.e. just declaring the plugins in the application's cargo.toml and calling them explicitly in the main() operation).

**Use case:**
The [zenoh-dds-plugin](https://github.com/eclipse-zenoh/zenoh-plugin-dds) can be built as a standalone zenoh application named `zenoh-bridge-dds`.  
We would like to administrate it at runtime in the same way as `zenohd`: via an admin space accessible through the REST API. The REST API is implemented as a [zenoh plugin](https://github.com/eclipse-zenoh/zenoh/tree/master/plugins/zenoh-rest). But we don't want the `zenoh-bridge-dds` to have the ability to dynamically load any plugin. The solution is to statically link `zenoh-bridge-dds` with the zenoh-rest plugin.
And as the code of the DDS/zenoh bridge itself should eventually be also a zenoh plugin (to be loadable by `zenohd`), the `zenoh-bridge-dds` will finally just by a main program starting the 2 plugins that are linked in static.

**Comment by JEnoch** [comment]
The current design using 2 `#[no_mangle]` function cannot work with statically linked plugins: the declaration of those functions by different plugins override each other, leading only 1 plugin to be accessible by the main().

Conditional compilation of `#[no_mangle]` attribute depending on the crate type (`cdylib` or `rlib`) doesn't work either, as long as rust-lang/rust#20267 is not fixed.

The most appropriate solution seems to follow the design suggested in https://michael-f-bryan.github.io/rust-ffi-guide/dynamic_loading.html:
 - define a trait to be implemented by each plugin struct
 - define a macro to be used by each plugin implementation to generate a "standardized" `#[no_mangle]` function to instantiate the plugin struct
 - the plugin manager in `zenohd` will call the `#[no_mangle]` function for each plugin
 - a standalone application which statically links with the plugins will directly instantiate and drive each plugin struct

**Comment by JEnoch** [comment]
Implemented by #114
